src/dataflow/analysis/analysis.ts

```diff
@@ -10,67 +10,229 @@
- * negated by "not" claims for the same tag downstream in the path, and ignoring
- * "not" claims without corresponding positive claims upstream, as these dangling
- * "not" claims are irrelevant for the given path. Note that "derives from" claims
- * only prune paths, and are dealt with during path generation, so they are ignored.
- */
-function computeFlowForPath(path: BackwardsPath): Flow {
-  const flow = new Flow();
-  for (const edge of path.edgesInForwardDirection()) {
-    if (edge.modifier) {
-      flow.modify(edge.modifier);
+class Solver {
+  readonly graph: FlowGraph;
+
+  /** Maps from an edge to a "expression" for it. */
+  readonly edgeExpressions: Map<Edge, EdgeExpression> = new Map();
+
+  /** Maps from an edge to the set of edges which depends upon it. */
+  readonly dependentEdges: Map<Edge, Set<Edge>>;
+
+  private _isSolved = false;
+
+  constructor(graph: FlowGraph) {
+    this.graph = graph;
+
+    // Fill dependentEdges map with empty sets.
+    this.dependentEdges = new Map();
+    for (const edge of graph.edges) {
+      this.dependentEdges.set(edge, new Set());
+    }
+  }
+
+  get isSolved() {
+    return this._isSolved;
+  }
+
+  validateAllChecks(): ValidationResult {
+    assert(this._isSolved, 'Graph must be solved before checks can be validated.');
+
+    const finalResult = new ValidationResult();
+    for (const edge of this.graph.edges) {
+      if (edge.check) {
+        const result = this.validateCheckOnEdge(edge);
+        result.failures.forEach(f => finalResult.failures.add(f));
+      }
+    }
+    return finalResult;
+  }
+
+  validateCheckOnEdge(edge: Edge): ValidationResult {
+    assert(this._isSolved, 'Graph must be solved before checks can be validated.');
+    assert(edge.check, 'Edge does not have any check conditions.');
+
+    const check = edge.check;
+    const finalResult = new ValidationResult();
+
+    const edgeExpression = this.edgeExpressions.get(edge);
+    const flows = edgeExpression.resolvedFlows;
+
+    for (const flow of flows) {
+      if (!flow.evaluateCheck(check)) {
+        // TODO: Figure out how to report the path that caused the failure.
+        finalResult.failures.add(`'${check.originalCheck.toManifestString()}' failed`);
+      }
+    }
+
+    return finalResult;
+  }
+
+  solve() {
+    assert(!this._isSolved, 'Graph is already solved.');
+
+    for (const edge of this.graph.edges) {
+      this.solveEdge(edge);
+    }
+
+    // Verify that all edges are fully resolved.
+    const numEdges = this.graph.edges.length;
+    assert(this.edgeExpressions.size === numEdges);
+    assert(this.dependentEdges.size === numEdges);
+    for (const edgeExpression of this.edgeExpressions.values()) {
+      assert(edgeExpression.isResolved, `Unresolved edge expression: ${edgeExpression.toString()}`);
+    }
+
+    this._isSolved = true;
+  }
+
+  solveEdge(edge: Edge) {
+    assert(!this.edgeExpressions.has(edge), 'Edge has already been solved.');
```

Comment:
Again, I don't think you need to assert here. Just return silently.

Reply:
Done